### DISCUSSION TITLE
Avatar upload (SvelteKit) unsuccessful

### DISCUSSION THREAD

**Discussion by YoRHa-A5** [pocketbase/pocketbase#1854] [Q&A] [answered]
Backend: Local PB instance connected to my S3 bucket
SvelteKit: 1.5.0
SkeletonUI: 0.124.2

Using SvelteKit formactions to receive the data, working fine so far. https://www.skeleton.dev/components/file-buttons is the component for the file input.

File arrives as
```ts
const data = await request.formData();
const avatar = data.get('avatar');

{
  name: 'xyz.jpg',
  lastModified: 1676400717302,
  type: 'image/jpeg',
  size: 174140,
  Symbol(kHandle): Blob {},
  Symbol(kLength): 174140,
  Symbol(kType): 'image/jpeg'
}
```

Also seems to be in the payload as `avatar: (binary)` and the request connects to the backend. All data except for the file gets written to the db. However, no avatar in the backend and no entry in the S3 bucket.

Am I missing something here? 🤔

**Comment by ganigeorgiev** [comment]
Are you sending the body as `FormData` object?

Are you sure that you are using the correct file field name in your request payload?

Could you provide a code sample how are you constructing and sending the request (or better a minimal reproducible repo)?

**Reply by ganigeorgiev** [reply]
_You can also check https://github.com/pocketbase/pocketbase/discussions/278#discussioncomment-3356869_

**Reply by YoRHa-A5** [reply]
Thanks, I will try with a new FormData object, tho typing and appending Arrays seems to be annoying here ^^
If it won't work I might set up a stackblitz or sth.

**Reply by ganigeorgiev** [reply]
If it doesn't work with manually constructing the `FormData`, you can also try the suggested answer from https://github.com/pocketbase/pocketbase/discussions/278#discussioncomment-3356869. Based on the comments in the linked issue, it is possible that it is an upstream issue in undici used by SvelteKit - https://github.com/sveltejs/kit/issues/5710#issuecomment-1195379338 (_although based on PR https://github.com/sveltejs/kit/pull/7065 it should have been already resolved_).

**Reply by YoRHa-A5** [reply]
> If it doesn't work with manually constructing the `FormData`, you can also try the suggested answer from [#278 (comment)](https://github.com/pocketbase/pocketbase/discussions/278#discussioncomment-3356869). Based on the comments in the linked issue, it is possible that it is an upstream issue in undici used by SvelteKit - [sveltejs/kit#5710 (comment)](https://github.com/sveltejs/kit/issues/5710#issuecomment-1195379338) (_although based on PR [sveltejs/kit#7065](https://github.com/sveltejs/kit/pull/7065) it should has been already resolved_).

Hm, weird that PR in Sveltekit was a long time ago so it should be fixed? 🤔
Anyway, here is what I currently do in my actions, maybe it helps some more?

```ts
export const actions: Actions = {
	setup: async ({ locals, request }) => {
		const data = await request.formData();
		const username = data.get('username');
		const about = data.get('about');
		const avatar = data.get('avatar');

		if (locals.user) {
			try {
				await locals.pb.collection('users').update(locals.user.id, {
					username: username,
					about: about,
					avatar: avatar
				});
			} catch (err) {
				if ((err as ClientResponseError).data?.code === 400) {
					return fail(400, { message: (err as ClientResponseError).data.message });
				} else if ((err as ClientResponseError).data?.code === 403) {
					return fail(403, { message: (err as ClientResponseError).data.message });
				} else if ((err as ClientResponseError).data?.code === 404) {
					return error(404, err as App.Error);
				} else {
					return error(400, err as App.Error);
				}
			}
			throw redirect(303, '/');
		}
	}
};
```

(If I need to do the node-fetch workaround, how do I import it so it gets used by the action?)

**Comment by ganigeorgiev** [comment, marked as answer]
@BunnyTheLifeguard In the above you are sending the body as json, but when uploading files the body must be a `FormData` object.

You can try either to send directly the `data` (if it is already `FormData`) or construct it on your own:

```js
await locals.pb.collection('users').update(locals.user.id, data);

// OR manually construct a new FormData object

const updateData = new FormData();
updateData.append("about", about);
updateData.append("avatar", avatar); // must be File or Blob
await locals.pb.collection('users').update(locals.user.id, updateData);
```

If the above doesn't work, you can try the workaround from https://github.com/sveltejs/kit/issues/5710#issuecomment-1258070291. If you want to use the SDK with node-fetch, you can reasign it to the global context (and optionally revert it after the action if you want the above to work only for the specific action), something like:
```js
import nodeFetch from 'node-fetch';

// or globalThis.fetch = nodeFetch; 
global.fetch = nodeFetch;
```

**Reply by YoRHa-A5** [reply]
Haha, no way! This makes form actions so much easier with Pocketbase, no `data.get()` anymore! 😀
I just needed to rename some input fields to match the backend fields/Ids and works like a charm now.

Thanks for your hard work, brother!
(Would love to contribute, but is hard to find time reading and understanding your code while working fulltime 😓)

**Comment by Tobfather** [comment]
Thank you!! omg, same happend to me! I was about to go crazy and finally found this thread!
